Ft api social login 167313401

Anguandia opened the pull request:
#### What does this PR do?

  Add the social login feature

#### Description of Task to be completed?

  - Login from social account if user exists
  - Create user and login if social user does not exist

#### How should this be manually tested?

  - If in a local environment, clone and install this repo locally and start the server with nom start/run dev common
  - In a browser enter the url http://:base-url/login/:social-account
    - The base-url is localhost:3000 for local testing and https://codepirates-ah-backend.herokuapp.com/ for remote testing
    - Social-account is facebook or google or twitter for the respective logins
  - In the resulting window; [facebook](https://user-images.githubusercontent.com/36167006/62922655-3869e780-bdac-11e9-8931-ee307bc5c380.png), [google](https://user-images.githubusercontent.com/36167006/62927228-6522fc80-bdb6-11e9-91a1-b5d6ff52d046.png), [twitter](https://user-images.githubusercontent.com/36167006/62923881-102fb800-bdaf-11e9-81c4-ef976d9d132d.png), enter your social media account credentials
  - If the account does not exist on author's haven, you will receive a [response](https://user-images.githubusercontent.com/36167006/62924005-74eb1280-bdaf-11e9-9bb4-f6491b6812eb.png) to consent to creating an account with Author's Haven
    - Enter localhost:3000/signup/social in the address bar and hit enter, you will see the successful signup message in the [response page](https://user-images.githubusercontent.com/36167006/62924131-b7acea80-bdaf-11e9-98e2-9a1b8b593412.png)
    - Open your associated email and confirm the address
    - Continue with theses steps
  - If the user account already exists on Author's Haven, you will  receive the [login success](https://user-images.githubusercontent.com/36167006/62922821-a57d7d00-bdac-11e9-891a-77a3678b9877.png) response containing your particulars and a token


#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?

  [finishes #167313401](https://www.pivotaltracker.com/story/show/167313401)

#### Screenshots (if appropriate)


facebook login page after entering localhost:3000/login/facebook in browser address bar in incognito window

<img width="1061" alt="facebook login page on virgin browser" src="https://user-images.githubusercontent.com/36167006/62922655-3869e780-bdac-11e9-8931-ee307bc5c380.png">


Alternative on personalised browser

<img width="1103" alt="possible facebook interface on given browser" src="https://user-images.githubusercontent.com/36167006/62922702-53d4f280-bdac-11e9-8bd8-3f8a24688951.png">


google login window on incognito window after entering localhost:3000/login/google in browser address bar

<img width="1396" alt="Screen Shot 2019-08-13 at 10 36 27" src="https://user-images.githubusercontent.com/36167006/62927228-6522fc80-bdb6-11e9-91a1-b5d6ff52d046.png">


Google login on personalised browser after entering localhost:3000/login/google in browser address bar

<img width="1125" alt="google login" src="https://user-images.githubusercontent.com/36167006/62922765-79fa9280-bdac-11e9-8208-2ef0474548bc.png">


Twitter login window after entering localhost:3000/login/twitter in browser address bar

<img width="936" alt="Screen Shot 2019-08-13 at 09 43 55" src="https://user-images.githubusercontent.com/36167006/62923881-102fb800-bdaf-11e9-81c4-ef976d9d132d.png">


Response prompting non existing social user consent for signup

<img width="1423" alt="Screen Shot 2019-08-13 at 09 47 30" src="https://user-images.githubusercontent.com/36167006/62924005-74eb1280-bdaf-11e9-9bb4-f6491b6812eb.png">


Response window on successful signup of social user

<img width="1425" alt="Screen Shot 2019-08-13 at 09 49 37" src="https://user-images.githubusercontent.com/36167006/62924131-b7acea80-bdaf-11e9-98e2-9a1b8b593412.png">


successful login response

<img width="1178" alt="Successful social login response" src="https://user-images.githubusercontent.com/36167006/62922821-a57d7d00-bdac-11e9-891a-77a3678b9877.png">

#### Questions:

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
feature(social login): user has to login with social platforms [Finishes #167313401]

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this PR

  Add social login

Tasks to be acomplished

  - Login from social account if user exists
  - Create user and login if social user does not exist

How to test manually

  - In a browser enter the url http://<base url>/login/<social account>
    - Get the base URL in the documentation
    - social account is facebook or google for the respective logins
  - Follow the prompts and observe the response in the browser

Related PT story

  [WIP #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

Anguandia committed:
feature(api-social-login):remove twitter [#167313401

habinezadalvan commented on src/controllers/socialLogin.js:
```diff
@@ -0,0 +1,62 @@
+  const dbSchema = {
+    firstname, lastname, email, username, password: hasspassword
+  };
+  const createdUser = await User.addUser(dbSchema);
```
@Anguandia So far you are on good progress. However, when a user is using social login and the user is not in Authors' heaven database he immediately signs up without making sure that he wants to sign up to Authors' heaven. I think it could be better if we first ask him if he is willing to proceed with sign up.

Anguandia replied:
Factored in, thanks for the observation

Anguandia committed:
Purpose

  Add social login feature

Tasks to be accomplished

  - Enable social login via twitter, google and facebook
  - Add creation of new users from social accounts on login
  - Document routes

How can this feature be tested manualy

  - In the browser address bar, enter the url https://codepirates-ah-backend.herokuapp.com/login/<social>,
where social can be facebook, google or twitter
  - In the resulting window, enter your credentials for the given social platform
  - If your account exists with this app, you should be logged in and your details and a token displayed in the browser
  - If your account does not exist, you will get a message in the broswer asking to creat an account
  - Enter https://codepirates-ah-backend.herokuapp.com/signup/social in the browser address bar and hit enter
  - You will get a message that your account has been created and a verification email sent along with your details and an access token

Pivotal tracker story

  [finishes #167313401]

Anguandia committed:
Purpose

  Add social login feature

Tasks to be accomplished

  - Enable social login via twitter, google and facebook
  - Add creation of new users from social accounts on login
  - Document routes

How can this feature be tested manualy

  - In the browser address bar, enter the url https://codepirates-ah-backend.herokuapp.com/login/<social>,
where social can be facebook, google or twitter
  - In the resulting window, enter your credentials for the given social platform
  - If your account exists with this app, you should be logged in and your details and a token displayed in the browser
  - If your account does not exist, you will get a message in the broswer asking to creat an account
  - Enter https://codepirates-ah-backend.herokuapp.com/signup/social in the browser address bar and hit enter
  - You will get a message that your account has been created and a verification email sent along with your details and an access token

Pivotal tracker story

  [finishes #167313401]

Anguandia committed:
disable failing tests for further inspection

Anguandia committed:
disable failing tests for further inspection

Mnickii commented:
Would you add screenshots to this? Is it still Work In Progress?

Anguandia committed:
resolve merge conflict

Anguandia committed:
resolve merge conflict

Anguandia committed:
Purpose of this commit:

  Resolve build failure

Story ID

  [finished #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
Purpose of this commit:

  Resolve build failure

Story ID

  [finished #167313401](https://www.pivotaltracker.com/story/show/167313401)

Anguandia committed:
fix history packag-lock error

Anguandia committed:
fix history packag-lock error

Anguandia committed:
work around twitter account security concern

Anguandia committed:
work around twitter account security concern

Anguandia committed:
work around twitter account security concern

Anguandia committed:
work around twitter account security concern

Anguandia committed:
pp

Anguandia committed:
pp

Anguandia committed:
twitter security

Anguandia committed:
twitter security

Anguandia committed:
twitter security

Anguandia committed:
twitter security

Anguandia committed:
twitter security

Anguandia committed:
twitter security

Anguandia committed:
Update test-socialLogin.js

Anguandia committed:
Update test-socialLogin.js

minega25 commented on src/controllers/social.js:
```diff
@@ -0,0 +1,100 @@
+ * @class Social
+ */
+class Social {
+  // constructor(data = {}) {
```
I think unused code should be removed

Anguandia replied:
Thanks Patrick, sorry forgot to remove commented-out code, let me work on these right away

minega25 commented on src/routes/api/index.route.js:
```diff
@@ -3,10 +3,13 @@ import swaggerui from 'swagger-ui-express';
 
 const router = express.Router();
 
+// router.use('/login', oauth);
```
same thing here as well. Remove unused code.

minega25 commented on src/routes/index.js:
```diff
@@ -12,14 +15,24 @@ dotenv.config();
 
 router.get('/', (req, res) => res.status(200).json({ status: 200, data: 'Welcome to Authors Haven.' }));
 router.use(baseUrl, api);
+router.use(oauth);
```
In my opinion, I think we should have our `api base url` before all routes in our API

Anguandia replied:
Let me harmonize

habinezadalvan reviewed (changes requested):
@Anguandia  I love the work you did the whole weekend, it seems you didn't sleep. Again I appreciate the fact that you worked on the feedback I gave you last time about asking the user if he wants to sign up to the Authors' heaven in case he doesn't have an account. However, as I was reviewing your work I realized that I can not log in with facebook; I think you have to give everyone access on facebook.
Check the screenshot below. Otherwise, you have my approval once that is done. Thank you.

![Screen Shot 2019-08-13 at 08 02 46](https://user-images.githubusercontent.com/45101716/62919105-ecb24080-bda1-11e9-8f02-8adae8edda22.png)

minega25 commented on test/users.test.js:
```diff
@@ -1,12 +1,7 @@
-import { verify } from 'crypto';
 
 let adminToken;
-// const adminToken = Helper.generateToken({
```
You can also remove this code as well.

minega25 reviewed (changes requested):
I think this is a job well done. Apart from some minor changes to make our codebase adhere to standards

habinezadalvan reviewed (commented):
@Could you please add more descriptions on how this feature can be tested locally. Please give the Url(s) for login and sign up in case the user doesn't exist on every social login. Eg: http://localhost:3000/login/facebook. Just specify every URL you think is important to us. Otherwise, I appreciate your work. Thank you.

nshutijonathan reviewed (commented):
Good work so far, however, the tests are not passing online We need to fix this

nkalyesubula reviewed (changes requested):
It looks good to me. The PR is failing because this repo on your side, is still tested under travis.org. Please switch to the travis-ci platform and push again the changes. Also, address the previous comments i.e add screenshots for twitter login, and remove comments.

Anguandia commented on test/verify.email.test.js:
```diff
@@ -0,0 +1,50 @@
+import 'dotenv/config';
```
Thanks for the observation, implemented

salviosage reviewed (commented):
Nice work Mike, LGTM if the previous comment issue is solved

Mnickii commented on .coveralls.yml:
```diff
@@ -1,2 +1 @@
-service_name: travis-pro
-repo_token: S5wTeingaGKIID1VojIa9l4McqoCl4YAi
\ No newline at end of file
+repo_token: process.env.COVERALLS_REPO_TOKEN
```
This is not part of your work. This is already merged. Please rebase using the current state of develop.

Mnickii commented on app.json:
```diff
@@ -43,6 +34,9 @@
       "STATE": {
         "required": true
+      },
+      "USER_PASSWORD": {
```
please rebase this branch well, looks like some history is messed up.

Anguandia commented:
@habinezadalvan the facebook app account is still a development account I am upgrading to production so everyone has access

Mnickii reviewed (changes requested):
Great work Mike. 

Please make sure this PR is an accurate depiction of only the changes you have worked on. You can achieve this by correctly rebasing the up to date develop branch.

Anguandia committed:
respond to feedback

Anguandia committed:
fix merge error

habinezadalvan reviewed (approved):
Great work @Anguandia;

Anguandia committed:
feature(social auth): add social authentication functionality

Mnickii reviewed (approved):
LGTM

salviosage commented:
LGTM
